src/connectivity.jl

```diff
@@ -129,9 +135,69 @@ julia> is_connected(g)
 true
 ```
 """
-function is_connected(g::AbstractGraph)
+function is_connected(g::AbstractGraph{T}) where {T}
     mult = is_directed(g) ? 2 : 1
-    return mult * ne(g) + 1 >= nv(g) && length(connected_components(g)) == 1
+    if mult * ne(g) + 1 >= nv(g)
+        label = zeros(T, nv(g))
+        connected_components!(label, g)
+        return allequal(label)
+    else
+        return false
+    end
+end
+
+"""
+    count_connected_components( g, [label, search_queue]; reset_label::Bool=false)
+
+Return the number of connected components in `g`.
+
+Equivalent to `length(connected_components(g))` but uses fewer allocations by not
+materializing the component vectors explicitly. Additionally, mutated work-arrays `label`
+and `search_queue` can be provided to reduce allocations further (see
+[`connected_components!`](@ref)).
+
+## Keyword arguments
+- `reset_label :: Bool` (default, `false`): if `true`, `label` is reset to zero before
+  returning.
+
+## Example
+```
+julia> using Graphs
+
+julia> g = Graph(Edge.([1=>2, 2=>3, 3=>1, 4=>5, 5=>6, 6=>4, 7=>8]));
+
+length> connected_components(g)
+3-element Vector{Vector{Int64}}:
+ [1, 2, 3]
+ [4, 5, 6]
+ [7, 8]
+
+julia> count_connected_components(g)
+3
+```
+"""
+function count_connected_components(
```

Comment:
I am a bit undecided if we should call this `count_connected_components` or `num_connected_components`. Currently we have both conventions, namely `num_self_loops` and `Graphs.Experimental.count_isomorph`.

Ideally we use the same word everywhere. @gdalle Do you have an opinion on that?

Reply:
There's also `nv(g)` for the number of vertices. Maybe just `nconnected_components`?

Reply:
If I had to pick I'd rather use `count` than `num` or `n` because it is a complete word

Reply:
Definitely no to `nconnected_components` - `nv` and `ne` might be some exceptions as they are used all the time - but we might rename them one day.

I don't mind abbreviation from time to time, but lets go with `count_connected_components` then - after all we also have a `count` function in the Julia base.